dot/rpc/modules/state.go

```diff
@@ -403,30 +406,72 @@ func (sm *StateModule) GetStorageSize(
 	return nil
 }
 
-// QueryStorage isn't implemented properly yet.
+// QueryStorage Query historical storage entries (by key) starting from a request start block given,
+//until request end block or best block if end block in nil
 func (sm *StateModule) QueryStorage(
 	_ *http.Request, req *StateStorageQueryRangeRequest, res *[]StorageChangeSetResponse) error {
 	if req.StartBlock.IsEmpty() {
-		return errors.New("the start block hash cannot be an empty value")
+		return ErrStartBlockHashEmpty
 	}
 
-	changesByBlock, err := sm.coreAPI.QueryStorage(req.StartBlock, req.EndBlock, req.Keys...)
+	startBlock, err := sm.blockAPI.GetBlockByHash(req.StartBlock)
 	if err != nil {
 		return err
 	}
 
-	response := make([]StorageChangeSetResponse, 0, len(changesByBlock))
+	startBlockNumber := startBlock.Header.Number
 
-	for block, c := range changesByBlock {
-		var changes [][]string
+	var endBlockHash common.Hash
+	if req.EndBlock.IsEmpty() {
+		endBlockHash = sm.blockAPI.BestBlockHash()
+	} else {
+		endBlockHash = req.EndBlock
+	}
+	endBlock, err := sm.blockAPI.GetBlockByHash(endBlockHash)
+	if err != nil {
+		return fmt.Errorf("cannot get block by hash: %w", err)
+	}
+	endBlockNumber := endBlock.Header.Number
 
-		for key, value := range c {
-			changes = append(changes, []string{key, value})
+	response := make([]StorageChangeSetResponse, 0)
+	lastValue := make([]*string, len(req.Keys))
+	firstPass := true
+	for i := startBlockNumber; i <= endBlockNumber; i++ {
+		blockHash, err := sm.blockAPI.GetHashByNumber(i)
+		if err != nil {
+			return fmt.Errorf("cannot get hash by number: %w", err)
 		}
+		var changes [][2]*string
+
+		for j, key := range req.Keys {
+			value, err := sm.storageAPI.GetStorageByBlockHash(&blockHash, common.MustHexToBytes(key))
+			if err != nil {
+				return fmt.Errorf("getting value by block hash: %w", err)
+			}
+			var hexValue *string
+			if len(value) > 0 {
+				hexValue = new(string)
+				*hexValue = common.BytesToHex(value)
+			}
 
-		blochHash := block
+			if firstPass {
+				changes = append(changes, [2]*string{stringPtr(key), hexValue})
+				lastValue[j] = hexValue
+				continue
+			}
+			if lastValue[j] == nil && hexValue != nil {
+				changes = append(changes, [2]*string{stringPtr(key), hexValue})
+				lastValue[j] = hexValue
+			}
+			if lastValue[j] != nil && *lastValue[j] != *hexValue {
+				changes = append(changes, [2]*string{stringPtr(key), hexValue})
+				lastValue[j] = hexValue
+			}
```

Comment:
```
			var hexValue string
			if len(value) > 0 {
				hexValue = common.BytesToHex(value)
			}

			if firstPass {
				changes = append(changes, [2]*string{stringPtr(key), &hexValue})
				lastValue[j] = &hexValue
				continue
			}
			if lastValue[j] == nil && &hexValue != nil {
				changes = append(changes, [2]*string{stringPtr(key), &hexValue})
				lastValue[j] = &hexValue
			}
			if lastValue[j] != nil && *lastValue[j] != hexValue {
				changes = append(changes, [2]*string{stringPtr(key), &hexValue})
				lastValue[j] = &hexValue
			}
```

Reply:
I know that having a string pointer is strange, but I did this so that the results for nil values show as JSON null.
Here is example output from polkadot v0.9.19 dev chain:
```
[
                    "0x1cb6f36e027abb2091cfb5110ab5087f06155b3cd9a8c9e5e9a23fd5dc13a5ed",
                    null
                ]
```
when I change and test gossamer (with runtime 0.9.19) with above changes I get output:
```
               [
                    "0x1cb6f36e027abb2091cfb5110ab5087f06155b3cd9a8c9e5e9a23fd5dc13a5ed",
                    ""
                ],
```
however when hexValue in a string pointer I get the same results as the polkadot node.  

Let me know if there is a different way to handle this so that the JSON parser treats nil values as `null` since I'd like to keep the output consistent with polkadot's output.

Reply:
removed test genesis spec `chain/dev/dev-spec-raw_gssmr9190.json`

Reply:
> I know that having a string pointer is strange, but I did this so that the results for nil values show as JSON null. Here is example output from polkadot v0.9.19 dev chain:
> 
> ```
> [
>                     "0x1cb6f36e027abb2091cfb5110ab5087f06155b3cd9a8c9e5e9a23fd5dc13a5ed",
>                     null
>                 ]
> ```
> 
> when I change and test gossamer (with runtime 0.9.19) with above changes I get output:
> 
> ```
>                [
>                     "0x1cb6f36e027abb2091cfb5110ab5087f06155b3cd9a8c9e5e9a23fd5dc13a5ed",
>                     ""
>                 ],
> ```
> 
> however when hexValue in a string pointer I get the same results as the polkadot node.
> 
> Let me know if there is a different way to handle this so that the JSON parser treats nil values as `null` since I'd like to keep the output consistent with polkadot's output.

Oh i see.  Yea it's a little awkward, but I see why you did that now.

Reply:
Although thinking about this, the empty string should be synonym with `null` right? (at least conceptually)
Why are they returning `null`? I guess we can just ignore this and return `null` too.